pandas/tests/arrays/test_datetimes.py

```diff
@@ -639,3 +639,21 @@ def test_tz_localize_t2d(self):
 
         roundtrip = expected.tz_localize("US/Pacific")
         tm.assert_datetime_array_equal(roundtrip, dta)
+
+    @pytest.mark.parametrize(
+        "error",
+        ["coerce", "raise"],
+    )
+    def test_coerce_fallback(self, error):
+        # GH#46071
```

Comment:
Thanks for adding this @srotondo. I don't think we're yet there, this doesn't seem to make so much sense.

My understanding is that with `infer_datetime_format=True` the format is inferred with the first non-NaN value, in this case `6/30/2025`, do the format should be `m/d/yyyy`. Then, when the next value is being parsed, it won't be able to, since `1 27 2024` is not a valid date in that format. When `errors='coerce'`, the returned Series should contain the first date, and the second `NaT`. When `errors='raise'` the `to_datetime` call should raise an exception.

This is very different to what you're asserting in this test, that it's that in both cases the result will be that both dates are being parsed.

Also, can you check if there is already a test for `to_datetime`? It should, it won't be testing this particular case, but maybe the same test can be reused in a way. Or at least have both tests next to each other.

Reply:
The original poster of this bug also thought that was how `infer_datetime_format` should work, but examining the programming closely makes it clear that it wants to avoid error returns as much as possible.

`infer_datetime_format` is meant to infer the first `datetime`'s format, and then cache that.  It will then attempt to use that inferred format on the subsequent `datetime`s to save time throughout the series.  If this format was made to be absolute for every term in the series, you would be correct that it should return the appropriate error result.  However, `infer_datetime_format` functions more like a guideline for the format.  There is then specific code in pandas that will fallback and try again to avoid errors for otherwise valid datetimes.  Raising an error or returning `NaT` should only happen if that fallback then proceeds to fail, for instance, if a datetime were out of range.

In terms of the test location, I could move this test to the `test_to_datetime.py` file, however, given that this test is mainly focused on the fallback itself and the piece of code that I changed wasn't directly in one of the `to_datetime` functions, but rather something it uses, I thought it would be more appropriate in the place I put it.  However, I'm completely open to moving it there.

Reply:
I guess that's the behavior we want to keep then. But we need to fix the documentation of `to_datetime`, to make clear what `infer_datetime_format` does. And we need to test properly anyway. If it only raises with a date out of range or that doesn't make sense, then we need to add a case for that to the test.